src/components/Note/Note.component.tsx

```diff
@@ -1,12 +1,153 @@
-import React from 'react';
+import React, { useState, useEffect, useRef } from 'react';
+import { BiArchiveIn, BiArchiveOut, BiEdit, BiSave } from 'react-icons/bi';
+import { RiDeleteBinLine } from 'react-icons/ri';
+import { VscSymbolColor } from 'react-icons/vsc';
+import { BlockPicker } from 'react-color';
 
-import { NoteWrapper } from './Note.styled';
+import { useNotesContext } from '../../providers/Notes';
+import useOnClickOutside from '../../hooks/useOnClickOutside';
+import { NoteProps } from '../../utils/types';
+import { NoteWrapper, NoteButton, SaveButton, ErrorAlert } from './Note.styled';
+
+function Note({ id, title, description, bgColor }: NoteProps) {
+  const {
+    updateNote,
+    deleteNote,
+    addArchiveNote,
+    removeArchiveNote,
+    isArchived,
+    updateArchivedNote,
+  } = useNotesContext();
+  const noteRef = useRef<HTMLDivElement | null>(null);
+  const titleRef = useRef<HTMLInputElement>(null!);
+  const [noteTitle, setNoteTitle] = useState(title);
+  const [noteDescription, setNoteDescription] = useState(description);
+  const [selectedColor, setSelectedColor] = useState(bgColor);
+  const [openColorPicker, setOpenColorPicker] = useState(false);
+  const [editMode, setEditMode] = useState(false);
+  const [error, setError] = useState('');
+
+  const selectedNote = { id, title, description, bgColor };
+
+  const handleUpdateNote = () => {
+    setError('');
+    const title = noteTitle;
+    const description = noteDescription;
+    const bgColor = selectedColor;
+
+    if (title.trim().length > 0 && description.trim().length > 0) {
+      if (isArchived(selectedNote)) {
+        updateArchivedNote({ id, title, description, bgColor });
+      } else {
+        updateNote({ id, title, description, bgColor });
+      }
+      setEditMode(false);
+    } else {
+      setError('The note cannot be empty');
+    }
+  };
+
+  function handleEditMode() {
+    setEditMode(true);
+  }
+
+  function handleDeleteNote() {
+    deleteNote(id);
+  }
+
+  function handleArchiveNote() {
+    addArchiveNote(selectedNote);
+  }
+
+  function handleUnarchiveNote() {
+    removeArchiveNote(selectedNote);
+  }
+
+  useOnClickOutside(noteRef, () => {
+    setEditMode(false);
+    setSelectedColor(bgColor);
+  });
+
+  useEffect(() => {
+    titleRef?.current?.focus();
+  }, [editMode]);
 
-function Note() {
   return (
-    <NoteWrapper>
-      <input placeholder="Title" type="text" required />
-      <textarea rows={3} placeholder="Type to add a note..."></textarea>
+    <NoteWrapper ref={noteRef} style={{ backgroundColor: selectedColor }}>
+      <div className="note-content">
+        {!editMode ? (
```

Comment:
What if we group all inputs using a React Frament <> ... </>. That way we can easily find the edit view and we can do the same for the read only version.

Reply:
Yes, I agree... this change has been completed!